core/server/services/auth/api-key/admin.js

```diff
@@ -34,89 +35,90 @@ const _extractTokenFromHeader = function extractTokenFromHeader(header) {
+    
+        const apiKeyId = decoded.header.kid;
+    
+        if (!apiKeyId) {
+            return next(new common.errors.BadRequestError({
+                message: common.i18n.t('errors.middleware.auth.adminApiKidMissing'),
+                code: 'MISSING_ADMIN_API_KID'
+            }));
+        }
+    
+        models.ApiKey.findOne({id: apiKeyId}).then((apiKey) => {
+            if (!apiKey) {
                 return next(new common.errors.UnauthorizedError({
-                    message: common.i18n.t('errors.middleware.auth.invalidTokenWithMessage', {message: err.message}),
-                    code: 'INVALID_JWT',
-                    err
+                    message: common.i18n.t('errors.middleware.auth.unknownAdminApiKey'),
+                    code: 'UNKNOWN_ADMIN_API_KEY'
                 }));
             }
-
-            // unknown error
-            return next(new common.errors.InternalServerError({err}));
-        }
-
-        // authenticated OK, store the api key on the request for later checks and logging
-        req.api_key = apiKey;
-        next();
-    }).catch((err) => {
-        next(new common.errors.InternalServerError({err}));
-    });
+    
+            if (apiKey.get('type') !== 'admin') {
+                return next(new common.errors.UnauthorizedError({
+                    message: common.i18n.t('errors.middleware.auth.invalidApiKeyType'),
+                    code: 'INVALID_API_KEY_TYPE'
+                }));
+            }
+    
+            // Decoding from hex and transforming into bytes is here to
+            // keep comparison of the bytes that are stored in the secret.
+            // Useful context:
+            // https://github.com/auth0/node-jsonwebtoken/issues/208#issuecomment-231861138
+            const secret = Buffer.from(apiKey.get('secret'), 'hex');
+    
+            // ensure the token was meant for this api
+            const options = Object.assign({
+                audience: urlUtils.getVersionPath({version, type: 'admin'})
+            }, JWT_OPTIONS);
+    
+            try {
+                jwt.verify(token, secret, options);
+            } catch (err) {
+                if (err.name === 'TokenExpiredError' || err.name === 'JsonWebTokenError') {
+                    return next(new common.errors.UnauthorizedError({
+                        message: common.i18n.t('errors.middleware.auth.invalidTokenWithMessage', {message: err.message}),
+                        code: 'INVALID_JWT',
+                        err
+                    }));
+                }
+    
+                // unknown error
+                return next(new common.errors.InternalServerError({err}));
+            }
+    
+            // authenticated OK, store the api key on the request for later checks and logging
+            req.api_key = apiKey;
+            next();
+        }).catch((err) => {
+            next(new common.errors.InternalServerError({err}));
+        });
+    }
```

Comment:
The original middleware fn is just wrapped into a method to capture `version` and used for audience calculation.